rectools/metrics/auc.py

```diff
@@ -0,0 +1,388 @@
+#  Copyright 2024 MTS (Mobile Telesystems)
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""ROC AUC based ranking recommendations metrics."""
+import typing as tp
+from enum import Enum
+
+import attr
+import pandas as pd
+
+from rectools import Columns
+from rectools.metrics.base import MetricAtK, outer_merge_reco
+from rectools.utils import select_by_type
+
+
+class InsufficientHandling(str, Enum):
+    """Strategy for handling insufficient reommendations cases"""
+
+    SKIP = "skip"
+    EXCLUDE = "exclude"
+    RAISE = "raise"
+
+
+@attr.s
+class AUCFitted:
+    """
+    Container with meta data got from `_AUCMetric.fit` method.
+
+    Parameters
+    ----------
+    outer_merged_enriched : pd.DataFrame
+        Recommendations outer merged with test interactions. Table has Columns.User, Columns.Item,
+        Columns.Rank. Precomputed columns include "__test_positive", "__tp", "__fp", "__fp_cumsum",
+        "__test_pos_cumcum". All ranks for all users are present with no skipping.
+        Null ranks are specified for test interactions that were not predicted in recommendations.
+    num_pos : pd.Series
+        Number of positive items for each user in test insteractions.
+    num_fp_insufficient : pd.Series
+        Number of false positive items for each user in `outer_merged_enriched` that had at least
+        one false negative. This users will be checked for insufficient cases processing.
+
+    """
+
+    outer_merged_enriched: pd.DataFrame = attr.ib()
+    num_pos: pd.Series = attr.ib()
+    num_fp_insufficient: pd.Series = attr.ib()
+
+
+class _AUCMetric(MetricAtK):
+    """
+    ROC AUC based metric base class.
+
+    Warning: This class should not be used directly.
+    Use derived classes instead.
+
+    Parameters
+    ----------
+    k : int
+        Number of items at the top of recommendations list that will be used to calculate metric.
+    insufficient_handling : {"skip", "raise", "exclude"}, default `"skip"`
```

Comment:
After a lot of thinking I decided that "raise" should not be the default option. It has a lot of logic that needs to be checked for correct "raise" behaviour. It requires a lot of understanding from user to be handled correctly. Key point is that not all models are compatible with "raise". I am very confident in this: we should do another option as default.

Reply:
Changed to 'ignore'